language-plutus-core/src/Language/PlutusCore/Type/Core.hs

```diff
@@ -0,0 +1,195 @@
+    | EqByteString
+    | LtByteString
+    | GtByteString
+    deriving (Show, Eq, Ord, Enum, Bounded, Generic, NFData, Lift)
+
+-- | The type of dynamic built-in functions. I.e. functions that exist on certain chains and do
+-- not exist on others. Each 'DynamicBuiltinName' has an associated type and operational semantics --
+-- this allows to type check and evaluate dynamic built-in names just like static ones.
+newtype DynamicBuiltinName = DynamicBuiltinName
+    { unDynamicBuiltinName :: Text  -- ^ The name of a dynamic built-in name.
+    } deriving (Show, Eq, Ord, Generic)
+      deriving newtype (NFData, Lift)
+
+-- | Either a 'BuiltinName' (known statically) or a 'DynamicBuiltinName' (known dynamically).
+data StagedBuiltinName
+    = StaticStagedBuiltinName  BuiltinName
+    | DynamicStagedBuiltinName DynamicBuiltinName
+    deriving (Show, Eq, Generic, NFData, Lift)
+
+data Builtin ann
+    = BuiltinName ann BuiltinName
+    | DynBuiltinName ann DynamicBuiltinName
+    deriving (Functor, Show, Generic, NFData, Lift)
+
+-- | A constant value.
+data Constant ann
+    = BuiltinInt ann Integer
+    | BuiltinBS ann BSL.ByteString
+    | BuiltinStr ann String
+    deriving (Functor, Show, Generic, NFData, Lift)
+
+data Term tyname name ann
+    = Var ann (name ann) -- ^ a named variable
+    | TyAbs ann (tyname ann) (Kind ann) (Term tyname name ann)
+    | LamAbs ann (name ann) (Type tyname ann) (Term tyname name ann)
+    | Apply ann (Term tyname name ann) (Term tyname name ann)
+    | Constant ann (Constant ann) -- ^ a constant term
+    | Builtin ann (Builtin ann)
+    | TyInst ann (Term tyname name ann) (Type tyname ann)
+    | Unwrap ann (Term tyname name ann)
+    | IWrap ann (Type tyname ann) (Type tyname ann) (Term tyname name ann)
+    | Error ann (Type tyname ann)
+    deriving (Functor, Show, Generic, NFData, Lift)
+
+type Value = Term
+
+-- | Version of Plutus Core to be used for the program.
+data Version ann
+    = Version ann Natural Natural Natural
+    deriving (Show, Functor, Generic, NFData, Lift)
+
+-- | A 'Program' is simply a 'Term' coupled with a 'Version' of the core language.
+data Program tyname name ann = Program ann (Version ann) (Term tyname name ann)
+    deriving (Show, Functor, Generic, NFData, Lift)
+
+newtype Normalized a = Normalized
+    { unNormalized :: a
+    } deriving (Show, Eq, Functor, Foldable, Traversable, Generic)
+      deriving newtype NFData
+      deriving Applicative via Identity
+deriving newtype instance PrettyBy config a => PrettyBy config (Normalized a)
+
+-- | All kinds of uniques an entity contains.
+type family HasUniques a :: Constraint
+type instance HasUniques (Kind ann) = ()
+type instance HasUniques (Type tyname ann) = HasUnique (tyname ann) TypeUnique
+type instance HasUniques (Term tyname name ann) =
+    (HasUnique (tyname ann) TypeUnique, HasUnique (name ann) TermUnique)
+type instance HasUniques (Program tyname name ann) = HasUniques (Term tyname name ann)
```

Comment:
I don't think this belongs to this module, but I'm not sure where else. Probably to some module related to the renamer. We'll settle on that in future.